### ISSUE TITLE
conan interactive mode even with CONAN_NON_INTERACTIVE

### ISSUE THREAD

**Issue by gunnwo** [conan-io/conan#4431]
All conan actions are done via script. This script creates a recipe and starts the download of packages from Artifactory. Because it's a script and not within a command prompt I've set the environment variable CONAN_NON_INTERACTIVE to true, always.
Previously I had stopped a download in the middle. A conan lock was left over from that break.
My expectation now was that with the existing lock and the non-interactive mode the conan command returns with an error and a proper error message. Instead the newly started download was on hold waiting for user interaction. Too bad that a script was doing the actions. It stopped until I had noticed the problem and killed the conan process.

Of course I can call "remove locks" before every operation. But sometimes this lock is happeing because of a parallel upload/download. 

This happens with older versions of conan as well as with the latest one. The only difference is the command "remove locks" which is now available.

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by memsharded** [comment]
I have had a look, and ran one test. The thing is:

- An uncontrolled kill of the Conan app can leave some package marked as locked. This is impossible to avoid.
- Any Conan process that tries to write over a locked package, will have to wait until that package is unlocked, to avoid concurrency issues.
- If the non-interactivity is enabled, failing with a message because a package is locked would basically break concurrency. That means that in non-interactive mode, nothing can be concurrent. It is not blocked for user input, it is actually blocked by the lock. But the lock could be there because of other Conan process, so it could be totally legit, no reason to fail.
- Also, doing a "remove locks" at the beginning of a script, can break other running Conan processes. It should be done if you are totally sure there are no other Conan processes running.

So I think there is nothing that conan could do, besides the remove locks command that now allows for an easier recovery of the situation.

**Comment by gunnwo** [comment]
Removing the locks globally is done only at the beginning of the script when there are no other up/downloads running. I will see how CONAN_USER_HOME will work. It could be that with the mass of the builds I'm running out of disk space with all the included third party. 
Of course there can be other uploads going on. My expectation was that the script will return if a lock exists, maybe not immediately but at least after some waiting time.

**Comment by memsharded** [comment]
> Removing the locks globally is done only at the beginning of the script when there are no other up/downloads running.

If you are sure there are no other ongoing transfers, you are good.

> Of course there can be other uploads going on. My expectation was that the script will return if a lock exists, maybe not immediately but at least after some waiting time.

But did you say that you know that there are no other transfers running? The problem of "waiting time" is twofold: first, there are many users build taking up to hours, so it is not possible to set a "reasonable" timeout of lets say, 5 minutes, and waiting time for that long is equivalent to having the build hanged.
Also, there are some locks that we could time-out, but others are just at the filesystem level, I am not sure if we could timeout them, without having to incur in some active-wait that would make the whole system much slower.

**Comment by memsharded** [comment]
Hi @gunnwo,

Reviewing again this issue, still think that nothing can be done, because of the reasons explained above in https://github.com/conan-io/conan/issues/4431#issuecomment-459299761. In summary:

- A lock is a lock, and has to be respected otherwise concurrency will be wildly broken. It is impossible to know if a lock is there because of an aborted run or because other process is running.
- Non interactive mode cannot disable or arbitrarily unlock the locks without breaking concurrency.
- Possible timeout unlocking is also arbitrary and can break concurrency. If the timeout is too small, it will break concurrency, if it is very large

The only possible ways I see are:
- Try to minimize concurrency issues using a cache per build with CONAN_USER_HOME
- Clearing possible stalled locks after an abort when you run your script and you know for sure no other concurrent process is accessing the cache.

So I think we can close this issue, but please ask any further question if necessary. Thanks for all the feedback!
